packages/docusaurus/src/server/utils.ts

```diff
@@ -31,3 +33,37 @@ export const logValidationBugReportHint = () => {
     )}\n`,
   );
 };
+
+// Recursively get the final routes (routes with no subroutes)
+export function getAllFinalRoutes(routeConfig: RouteConfig[]): RouteConfig[] {
+  function getFinalRoutes(route: RouteConfig): RouteConfig[] {
+    return route.routes ? flatMap(route.routes, getFinalRoutes) : [route];
+  }
+  return flatMap(routeConfig, getFinalRoutes);
+}
+
+// Report message to user through the console with varying severity
+export function reportMessage(
```

Comment:
I think this could be used for reporting things other than errors, so I named it in a general way. 
I prepended the type of message (info, warn, error) in front of the body so it looks similar to the other messages